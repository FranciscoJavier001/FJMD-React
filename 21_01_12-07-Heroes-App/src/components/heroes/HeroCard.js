//**_______________________________________________________________________________________________________________________________________________*/
import React from 'react';
import { Link } from 'react-router-dom';

export const HeroCard = ({ //** Esta es la informacion que esperamos recibir y la tenemos en data/heroes */
  id,
  superhero,
  alter_ego,
  first_appearance,
  characters,
}) => {

  return (

    <div className="card ms-3" style={{ maxWidth: 540 }}> {/* card=Que se muestren tarjetas, style, es la anchura maxima de las tarjetas */}
      <div className="row no-gutters"> {/* row=Que salgan en columnas(la info alado) no-gutters=Imagenes ocupen todo el largo de la tarjeta */}
        <div className="col-md-4"> {/* col=Columnas que ocupa la tarjeta 4 de 12 */}
          <img //** Para mostrar las imagenes */
            src={`./assets/heroes/${ id }.jpg`} //** Donde se Localizan para añadirlos */
            className="card-img" //** Para que ocupe todo el largo la imagen */
            alt={superhero} //** Para el nombre del superheroe */
          />
        </div>
        <div className="col-md-8"> {/* Espacio que ocupa la info, ya ocupe 4, solo quedan 8 */}
          <div className="card-body">
            <h5 className="card-title">{superhero}</h5> {/* h5=Encabeado 5(tamaño), card-title=Para poner el nombre del superheroe */}
            <p className="card-text">{alter_ego}</p> {/* p=parrafo, card-text=Texto del nombre del superheroe */}
            {/* { ( alter_ego !== characters ) //** alter_ego(nombre en la vida real) es diferente caracter(nombre de sus personajes) */
              // && <p className="card-text">{ characters }</p> //** && renderiza otro p, p=parrafo, card-text=Mostramos el nombre de sus personajes */
            }
            <p className="card-text"> {/* card-text=Texto en la tarjeta */}
              <small className="text-muted"> {first_appearance} </small> {/* small=texto chico, text-muted=texto en gris, historieta que aparecio */}
            </p>
            
            <Link to={ `./hero/${ id }`}> {/* Si le ponen en mas te lleva a un enlace del Heroe con su informacion components/heroes/HeroScreen*/}
                Más...
            </Link>
          </div>
        </div>
      </div>
    </div>
  )
}
