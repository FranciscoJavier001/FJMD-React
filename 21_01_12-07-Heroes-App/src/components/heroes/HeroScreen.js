//**_______________________________________________________________________________________________________________________________________________*/
import React, { useMemo } from 'react'
import { useParams, Redirect } from 'react-router-dom'
import { getHeroById } from '../../selectors/getHeroById';

export const HeroScreen = ({ history }) => { //** Necesito el history para poder modificar el boton (Puedo ver las props en console_compoents) */

    const { heroeId } = useParams(); //** useParams=Extrae parametros del url, lo obtenemos como params que desestructuramos con heroeId */
    // console.log(params); //** l8 tenia params, pero lo desestructuramos, hacemos esto viendo lo que trae en consola y lo puedo traer */
    
    // const hero = getHeroById( heroeId ); //** Funcion de getHeroById que manda el heroId de params"este manda, funcion de flecha recibe" */
    // console.log(hero); //** Vemos que tenemos toda la info del heroe */

    //** useMemo, memoriza para que no haya cambios en el componente, hasta que haya cambio en la dependencia lo segundo */
    //** En la funcion de flecha que se ejecuta, lo primero es la funcion a llamar l8, y lo segundo lo que se dispara si el heroe es diferente */
    const hero = useMemo(() => getHeroById( heroeId ), [ heroeId ]);
    
    // console.log('Hero:', hero, 'HeroId:' , heroeId,); //** Viene info del hero, y lo segundo es heroeId, que viene en data/heroes */
    
    if ( !hero ) { //** Si el heroe no existe voy a mandar el Redirect que viene de react-router-dom y lo redirecciono a la pagina de marvel*/
        return <Redirect to="/" />
    }

    const handleReturn = () => { //** Funcionalidad del boton */

        if ( history.length <= 2 ) { //** Si el historial de navegacion es menor o igual a 2 entonces llevame al /=marvel */
            history.push('/');
        } else {
            history.goBack(); //** Si es mayoro igual a 3, entonces solo llevame un historial hacia atras, pagina dc/marvel */
        }
    }
    
    //** Voy a desestructuras las propiedades del heroe y las voy a extraer en el hero, las tengo en data/heroes, solo las copie */
    const { 
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters,
    } = hero;

    return (
        //** Cuando le ponemos en mostrar mas, nos pone la imagen del superheroe en una nueva pagina */
        <div className="row mt-5"> {/* row=Que salga en filas(hacia la derecha) - mt-5=Distancia del margen superior */}
            <div className="col-4"> {/* Espacio que ocupa la imagen */}
                <img
                    src={ `../assets/heroes/${ heroeId }.jpg` } //** Path que se va a mostrar */
                    alt={ superhero } //** Lo que se muestra en caso que no se pueda mostrar la imagen */
                    //** img-thumbnail=Solo ocupe espacio asignado */
                    //** animate__animated=Activamos la opcion - animate__fadeInLeft=Desvanece desde la izquierda */
                    className="img-thumbnail animate__animated animate__fadeInLeft" 
                    />
            </div>

            {/* col-8=Espacio donde viene la info - animate__animated=activarlo - animate__fadeIn=Efecto aparicion como desvanecido */}
            <div className="col-8 animate__animated animate__fadeIn">
                <h3 className='text-center'><u>Superhero: { superhero } -  Characters: { characters } </u></h3> {/* Encabezado, u=subrayado */}
                <ul className="list-group list-group-flush"> {/* ui=Lista - list-group=En lista con margen - list-group-flush=Quitar ls tablas */}
                    <li className="list-group-item"> <b> Alter Ego: </b> { alter_ego } </li> {/* list-group-item=Quitar el simbolo de lista */}
                    <li className="list-group-item"> <b> Publisher: </b> { publisher } </li> {/* b=Negritas */}
                    <li className="list-group-item"> <b> First Appearance: </b> { first_appearance } </li>
                </ul>

                {/* <h5> Characters </h5> <p> { characters }</p> Encabezado */}
                {/* <p> { characters } </p> Se muestra abajo */}


                <button //** Que esto va a ser un boton */
                className="btn btn-outline-info" //** btn=Boton bootstrap - btn-outline-info=Boton transparente, con margen y azul */
                onClick={handleReturn} //** Accion que manda al hacer click */
                >
                    Return {/* Lo que dice */}
                </button>
            </div>
            
        </div>
    )
}