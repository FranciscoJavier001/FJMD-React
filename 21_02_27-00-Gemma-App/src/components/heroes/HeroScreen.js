import React, { useMemo } from 'react'
import { useParams, Redirect } from 'react-router-dom'
import { getHeroById } from '../../selectors/getHeroById';

export const HeroScreen = ({ history }) => {

    //** Si quiero extraer el HeroScreen hay un customHook que se llama useParams, este hook va a extraer los parametros que vallan por el url y los vamos a obtener como params, si necesitamos el heroeId pues vamos a desestructurar el heroeId */
    const { heroeId } = useParams();
    
    //** Ahora voy a sacar toda la informacion del hero y lo voy a sacar con el getHeroeById que esta en nuestros selectores y mandamos el heroe.Id*/
    //** Aqui tambien voy a utilizar el useMemo para que no haya cambios, si no hay cambios en el componente */
    const hero = useMemo(() => getHeroById( heroeId ), [ heroeId ]); //** Lo primero es la funcion que vamos a llamar y lo segundo es lo que se va a disparar cada vez que el heroeId sea diferente, y quite esta linea de abajo que es lo mismo, pero aqui con el useMemo "const hero = getHeroById( heroeId );" */
    // console.log('Hero: ' , hero, 'HeroId: ' , heroeId,); //** Lo removi para que no estuiera muy cargada de console.log las pruebas */
    
    //** Si el heroe no existe voy a mandar el Redirect que viene de react-router-dom y lo redirecciono a la pagina de marvel*/
    if ( !hero ) {
        return <Redirect to="/" />
    }

    //** Asi creamos funcionabilidades de los botones */
    const handleReturn = () => { //** No se recibe ningun argumento y este va a ser la funcion del boton */

        //** Asi revisamos el length para hacer una validacion, por lo que entiendo, el push solo me regresa un escalon, pero solo funciona si es mayor a 2 el historial y si es menor, entonces directamente me lleva al inicio, wey en las herramientas de desarrollo estan estas son desde el push, goback, goforward y demas */
        //** Si el history.length es menor o igual a 2 entonces quiero regresarme a la pagina principal de la aplicacion que seria el history.push y quiero que navege a la pagina de marvel, en caso contrario solo que navege atras */
        if ( history.length <=2 ) {
            history.push('/');
        } else {
            history.goBack();
        }
    }
    
    //** Voy a desestructuras las propiedades del heroe y las voy a extraer en el hero */
    const { 
        superhero,
        alter_ego,
        first_appearance,
        characters,
        tiempo,
    } = hero;

    return (
        //** Cuando le ponemos en mostrar mas, nos pone la imagen del superheroe en una nueva pagina */
        <div className="row mt-5">
            <div className="col-4">
                <img
                    src={ `../assets/heroes/${ heroeId }.jpeg` }
                    alt={ superhero } //** Lo que se muestra en caso que no se pueda mostrar la imagen */
                    className="img-thumbnail animate__animated animate__fadeInLeft"
                    />
            </div>

{/* Wey, esto ya es lo que esta dentro de las tarjetas, cuando le damos en mas */}
            <div className="col-8 animate__animated animate__fadeIn">
                <h3>{ superhero }</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"> <b> Producto: </b> { alter_ego } </li>
                    <li className="list-group-item"> <b> Tiempo: </b> { tiempo } </li>
                    <li className="list-group-item"> <b> Precio: </b> { first_appearance } </li>
                    <li className="list-group-item"><h5> Descripcion: </h5> { characters }</li>
                </ul>


                <button 
                className="btn btn-outline-info btn-group-justified mr-2"
                onClick={handleReturn}
                >
                    Regresar
                </button>


                <button
                className="btn btn-primary btn-group-justified mr-2 active"
                >
                    Reserva
                </button>

                {/* <button
                className="btn btn-danger btn-group-justified mr-2 disabled "
                >
                    Agotado
                </button> */}
            </div>
        </div>
    )
}
