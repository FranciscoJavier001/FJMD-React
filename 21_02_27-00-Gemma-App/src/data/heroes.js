export const heroes = [

    {
        'id': 'masaje-masaje_reductivo',
        'superhero':'Masaje Reductivo', 
        'publisher':'servicio', 
        'alter_ego':'10 Sesiones + Pedicura',
        'first_appearance':'$2,999.00',
        'characters':'Este es un Paquete para que te consientas, aqui se incluyen 10 sesiones de Masajes Reductivos y Pedicura, ', 
        'tiempo':'60 min'
    },
    {
        'id': 'pies-hongos_bye',
        'superhero':'Hongos Bye', 
        'publisher':'servicio', 
        'alter_ego':'Tratamiento para Hongos',
        'first_appearance':'$300.00',
        'characters':'Tratamiento para la eliminacion de Hongos',
        'tiempo':'45 min'
    },
    {
        'id': 'Masajes',
        'superhero':'Masajes', 
        'publisher':'servicio', 
        'alter_ego':'Masaje Rejanate',
        'first_appearance':'$300.00',
        'characters':'Masaje Relajante, que incluye Parches de Colageno de Hidratacion Extrema y una Bebida de Cortesia',
        'tiempo':'60 min'
    },
    // {
    //     'id': '',
    //     'superhero':'', 
    //     'publisher':'servicio', 
    //     'alter_ego':'',
    //     'first_appearance':'$',
    //     'characters':'',
    //     'tiempo':''
    // },
    
]