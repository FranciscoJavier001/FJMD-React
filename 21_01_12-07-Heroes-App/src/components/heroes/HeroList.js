//**_______________________________________________________________________________________________________________________________________________*/
import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher'
import { HeroCard } from './HeroCard';

export const HeroList = ({ publisher }) => { //** Recibimos publisher de src/data/heroes (es un atributo), de quien lo publico */
    
    //** Creo una const, que como no va a estar mutando la info la recibe de la funcion gHBP, recibiendo el publisher, asi tengo la data */
    //** const heroes = getHeroesByPublisher( publisher ); */

    //** useMemo, memoriza valores, recibe una funcion que retorna el valora a memorizar, segundo es lo que se dispara si hay cambios */
    const heroes = useMemo(() => getHeroesByPublisher( publisher ), [publisher])

    return (
        // Esto lo vamos a poner ahora como un div y le vamos a poner una clase de bootstrap
        //** No quiere detectar las columnas de bootstrap, hay que analizar que pasa */
        //** Esto ultimo que le puse es una clase de animate.css, donde copie el cdn en el index y ya solo use lo que queria de estilo */
        //** animate.css es un cdn importada en index */
        //** card-columns=Para ver todo el Columnas, pero siempre deja 2 solos si no esta completo */
        //** animate__animated=Para asignar la animacion */
        //** animate__fadeIn=Efecto como desvanecido luego aparezcan */
        <div className="card-columns animate__animated animate__fadeIn"> 
             { //** Esta es la expresion para barrer osea usar el map */
                heroes.map( hero => ( //** Barro el arreglo de los heroes con una funcion hero que retorna un objeto en el HC l25 */
                    <HeroCard //** Importado desde components/>heroes/HeroCard, y es para poder renderizar la tareta de los heroes */
                    key={ hero.id } //** Key es para acceder a un atributo de HC "id" de la constante exportada en l5 components/heroes/HeroCard */
                    { ...hero } //** Operador spred para mostrar la info de las propiedades de cada heroe */
                    />
                 ))
                }
        </div>
    )
}