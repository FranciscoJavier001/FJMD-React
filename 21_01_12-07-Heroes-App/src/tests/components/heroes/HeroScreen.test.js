//**_______________________________________________________________________________________________________________________________________________*/
import React from 'react'
import { shallow, mount } from 'enzyme'
import { HeroScreen } from '../../../components/heroes/HeroScreen';
import { MemoryRouter, Route } from 'react-router-dom';

describe('Pruebas en <HeroScreen />', () => {
    
    const history = { 
        length: 10, 
        push: jest.fn(), 
        goBack: jest.fn(), 
    }

    const wrapper = mount( 
        <MemoryRouter initialEntries={['/hero']}>
            <HeroScreen history={ history } />
        </MemoryRouter>
    );

    test('Debe de mostrar el componente redirect si no hay argumentos en el URL', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
            <HeroScreen history={ history } />
            </MemoryRouter>
        );
        
        expect( wrapper ).toMatchSnapshot()
        expect( wrapper.find('Redirect').exists() ).toBe(true);
    })

    test('Debe de mostrar un heroe si el parametro existe y se encuentra', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroeId" component={ HeroScreen } />
            </MemoryRouter>
        )

        expect( wrapper.find('.row').exists() ).toBe(true)
    })

    test('Debe de regresar a la pantalla anterior con PUSH', () => {
        
        const history = { 
        length: 1,
        push: jest.fn(),
        goBack: jest.fn(),
    }
    
    const wrapper = mount(
        <MemoryRouter initialEntries={['/hero/marvel-spider']}>
            <Route
                path="/hero/:heroeId" 
                component={ () => <HeroScreen history={ history } /> } 
            />
        </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')()

        expect( history.push ).toHaveBeenCalledWith('/')
        expect( history.goBack ).not.toHaveBeenCalled()
    })

    test('Debe de regresar a la pantalla anterior GOBACK', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route
                    path="/hero/:heroeId" 
                    component={ () => <HeroScreen history={ history } /> } 
                />
            </MemoryRouter>
            )
    
            wrapper.find('button').prop('onClick')()
    
            expect( history.push ).toHaveBeenCalledTimes(0) //** Que haya sido llamado 0 veces */
            expect( history.goBack ).toHaveBeenCalled() //** Que haya sido llamado */
    })

    test('debe de llamar el redirect si el heroe no existe', () => {
        

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider0']}>
                <Route
                    path="/hero/:heroeId" 
                    component={ () => <HeroScreen history={ history } /> } 
                />
            </MemoryRouter>
            )
    
            expect( wrapper.text() ).toBe('');
    })
})