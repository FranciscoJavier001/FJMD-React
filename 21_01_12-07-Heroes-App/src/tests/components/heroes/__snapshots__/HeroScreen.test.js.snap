// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen /> Debe de mostrar el componente redirect si no hay argumentos en el URL 1`] = `""`;
